internal/provider/data_source_http.go

```diff
@@ -95,12 +124,39 @@ func (d *httpDataSource) Read(ctx context.Context, req tfsdk.ReadDataSourceReque
 		return
 	}
 
+	var retry retryModel
+
+	if !model.Retry.Null && !model.Retry.Unknown {
+		diags = model.Retry.As(ctx, &retry, types.ObjectAsOptions{})
+		resp.Diagnostics.Append(diags...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
 	url := model.URL.Value
 	headers := model.RequestHeaders
+	timeout := model.RequestTimeout
+
+	var cancel context.CancelFunc
+
+	if !timeout.IsNull() {
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(timeout.Value)*time.Millisecond)
+		defer cancel()
+	}
 
-	client := &http.Client{}
+	retryClient := retryablehttp.NewClient()
+	retryClient.Logger = levelledLogger{ctx}
```

Comment:
Aside: This leaves us with an interesting inflection point with terraform-plugin-log. We recently updated the `http.Transport` in sdk/v2 helper/logging for terraform-plugin-log support: https://www.terraform.io/plugin/sdkv2/logging/http-transport

We thought that it might be the case that framework providers could use something similar and that the answer there might be copying that `http.Transport` implementation into a new terraform-plugin-log package, since its extending Go standard library functionality and tightly coupled with terraform-plugin-go functionality.

That said, the logging implementation of go-retryablehttp partially overlaps/conflicts with some of that `http.Transport` implementation if we were to try an also use it here. For example, retryablehttp saves some information in differently named fields and it would always create duplicate log entries for sending/receiving requests unless we overwrite `RequestLogHook`/`ResponseLogHook`. The terraform-plugin-log `http.Transport` implementation also creates a transaction ID for correlating entries easier.

I'm not sure if its worth looking into this further, but wanted to raise it in case it seems like an area we should explore more. I think go-retryablehttp is a really good choice in general since it bakes in a lot of goodies, just not sure about the logging side of things.

Reply:
Created https://github.com/hashicorp/terraform-plugin-log/issues/91 to discuss some ideas.